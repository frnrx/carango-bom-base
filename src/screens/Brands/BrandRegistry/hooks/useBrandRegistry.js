import { useState, useContext } from 'react';
import { useHistory } from 'react-router-dom';

import { SnackBarContext } from '../../../../contexts/snackbar';
import { AuthenticationContext } from '../../../../contexts/authentication';

import { registerBrand } from '../../services';

const useBrandRegistry = () => {
  const [isLoading, setIsLoading] = useState(false);
  const { addAlert } = useContext(SnackBarContext);
  const { userJWT } = useContext(AuthenticationContext);
  const history = useHistory();

  const register = (brand) => {
    setIsLoading(true);
    registerBrand(userJWT, brand)
      .then(() => {
        addAlert({ content: 'Marca cadastrada corretamente!', customSeverity: 'success' });
        history.push('/marcas');
      })
      .catch(() => {
        addAlert({ content: 'Erro inesperado ao cadastrar marca!', customSeverity: 'error' });
      })
      .finally(() => {
        setIsLoading(false);
      });
  };

  return { register, isLoading };
};

export default useBrandRegistry;