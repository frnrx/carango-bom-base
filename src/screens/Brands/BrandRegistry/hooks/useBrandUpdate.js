import { useState, useContext, useEffect } from 'react';
import { useHistory } from 'react-router-dom';

import { SnackBarContext } from '../../../../contexts/snackbar';
import { AuthenticationContext } from '../../../../contexts/authentication';

import { updateBrand, getBrand } from '../../services';

const useBrandUpdate = (brandId, setBrand) => {
  const [isUpdateLoading, setIsUpdateLoading] = useState(false);
  const [isGetLoading, setIsGetLoading] = useState(false);
  const { addAlert } = useContext(SnackBarContext);
  const { userJWT } = useContext(AuthenticationContext);
  const history = useHistory();

  const update = (brand) => {
    setIsUpdateLoading(true);
    updateBrand(userJWT, brandId, brand)
      .then(() => {
        addAlert({ content: 'Marca atualizada corretamente!', customSeverity: 'success' });
        history.push('/marcas');
      })
      .catch(() => {
        addAlert({ content: 'Erro inesperado ao atualizar marca!', customSeverity: 'error' });
      })
      .finally(() => {
        setIsUpdateLoading(false);
      });
  };

  const loadBrand = () => {
    setIsGetLoading(true);
    getBrand(brandId, userJWT)
      .then((data) => {
        setBrand(data.nome);
      })
      .catch(() => {
        addAlert({
          content: 'Erro inesperado ao carregar informações da marca!',
          customSeverity: 'error',
        });
        history.push('/marcas');
      })
      .finally(() => {
        setIsGetLoading(false);
      });
  };

  useEffect(() => {
    if (brandId) {
      loadBrand();
    }
  }, [brandId]);

  return { update, isUpdateLoading, isGetLoading };
};

export default useBrandUpdate;
