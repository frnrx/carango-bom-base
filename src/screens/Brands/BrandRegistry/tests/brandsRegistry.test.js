import React from 'react';
import { MemoryRouter } from 'react-router-dom';
import { render, screen, cleanup, fireEvent } from '@testing-library/react';

import BrandsRegistry from '..';
import useBrandRegistry from '../hooks/useBrandRegistry';
import useBrandUpdate from '../hooks/useBrandUpdate';

jest.mock('../hooks/useBrandRegistry');
jest.mock('../hooks/useBrandUpdate');

const mockedBrandId = 12;
let mockedLocationPathname = '/cadastro-marca';

jest.mock('react-router-dom', () => ({
  ...jest.requireActual('react-router-dom'),
  useLocation: () => ({
    pathname: mockedLocationPathname,
  }),
  useParams: () => ({
    brandId: mockedBrandId,
  }),
}));

describe('Create/update brand form', () => {
  let brandInput;
  let cancelLink;

  beforeEach(() => {
    useBrandRegistry.mockReturnValue({ register: jest.fn() });
    useBrandUpdate.mockReturnValue({ update: jest.fn() });
    render(
      <MemoryRouter>
        <BrandsRegistry />
      </MemoryRouter>,
    );

    brandInput = screen.getByRole('textbox', { name: 'Marca' });
    cancelLink = screen.getByRole('link', { name: 'Cancelar' });
  });

  it('should match the snapshot', () => {
    const { container } = render(
      <MemoryRouter>
        <BrandsRegistry />
      </MemoryRouter>,
    );
    expect(container.firstChild).toMatchSnapshot();
  });

  describe('render', () => {
    it('should render form default components correctly', () => {
      expect(brandInput).toBeInTheDocument();
      expect(cancelLink).toBeInTheDocument();
    });

    it('should render the registry button and heading when the registry route is used', () => {
      const registryHeading = screen.getByText('Cadastro de marcas');
      const registryButton = screen.getByRole('button', { name: 'Cadastrar' });
      expect(registryButton).toBeInTheDocument();
      expect(registryHeading).toBeInTheDocument();
    });

    it('should render the update button and heading when the update route is used', () => {
      cleanup();
      mockedLocationPathname = '/alteracao-marca';
      render(
        <MemoryRouter>
          <BrandsRegistry />
        </MemoryRouter>
      );
      const updateHeading = screen.getByText('Alteração de marca');
      const updateButton = screen.getByRole('button', { name: 'Alterar' })
      expect(updateButton).toBeInTheDocument();
      expect(updateHeading).toBeInTheDocument();
    });

    it('should render the loading component when the register is loading (registry mode)', () => {
      cleanup();
      mockedLocationPathname = '/cadastro-marca';
      useBrandRegistry.mockReturnValue({ register: jest.fn(), isLoading: true });
      useBrandUpdate.mockReturnValue({ update: jest.fn() });
      const { getByRole } = render(
        <MemoryRouter>
          <BrandsRegistry />
        </MemoryRouter>,
      );
      expect(getByRole('progressbar')).toBeInTheDocument();
    });

    it('should render the loading component when the update is loading (update mode)', () => {
      cleanup();
      mockedLocationPathname = '/alteracao-marca';
      useBrandRegistry.mockReturnValue({ register: jest.fn() });
      useBrandUpdate.mockReturnValue({ update: jest.fn(), isUpdateLoading: true });
      const { getByRole } = render(
        <MemoryRouter>
          <BrandsRegistry />
        </MemoryRouter>,
      );
      expect(getByRole('progressbar')).toBeInTheDocument();
    });

    it('should render the loading component when the get is loading (update mode)', () => {
      cleanup();
      mockedLocationPathname = '/alteracao-marca';
      useBrandRegistry.mockReturnValue({ register: jest.fn() });
      useBrandUpdate.mockReturnValue({ update: jest.fn(), isGetLoading: true });
      const { getByRole } = render(
        <MemoryRouter>
          <BrandsRegistry />
        </MemoryRouter>,
      );
      expect(getByRole('progressbar')).toBeInTheDocument();
    });
  });

  describe('form and validation functioning', () => {
    it('should send the data correctly to the update service when a user presses submit', () => {
      cleanup();
      const mockedUpdateService = jest.fn();
      mockedLocationPathname = `/alteracao-marca/${mockedBrandId}`;
      useBrandRegistry.mockReturnValue({ register: jest.fn() });
      useBrandUpdate.mockReturnValue({ update: mockedUpdateService });
      const { getByRole } = render(
        <MemoryRouter>
          <BrandsRegistry />
        </MemoryRouter>,
      );
      brandInput = getByRole('textbox', { name: 'Marca' });
  
      fireEvent.change(brandInput, { target: { value: 'Fake car brand' } });
      fireEvent.focusOut(brandInput);

      const expectedForm = {
        vehicleId: mockedBrandId,
        brand: 'Fake car brand',
      };

      const updateButton = getByRole('button', { name: 'Alterar' });
      fireEvent.click(updateButton);
  
      expect(mockedUpdateService).toHaveBeenCalledWith(expectedForm.brand);
    });

    it('should send the data correctly to the registry service when a user presses submit', () => {
      cleanup();
      const mockedRegisterVehicle = jest.fn();
      mockedLocationPathname = '/cadastro-marca';
      useBrandRegistry.mockReturnValue({ register: mockedRegisterVehicle });
      useBrandUpdate.mockReturnValue({ update: jest.fn() });
      const { getByRole } = render(
        <MemoryRouter>
          <BrandsRegistry />
        </MemoryRouter>,
      );
      brandInput = getByRole('textbox', { name: 'Marca' });
  
      const mockedBrandName = 'Fake brand name';
      fireEvent.change(brandInput, { target: { value: mockedBrandName } });
      fireEvent.focusOut(brandInput);

      const updateButton = getByRole('button', { name: 'Cadastrar' });
      fireEvent.click(updateButton);
  
      expect(mockedRegisterVehicle).toHaveBeenCalledWith(mockedBrandName);
    });
  });
});
