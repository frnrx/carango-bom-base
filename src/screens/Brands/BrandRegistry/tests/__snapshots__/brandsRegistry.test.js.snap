// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Create/update brand form should match the snapshot 1`] = `
<div
  class="MuiBox-root MuiBox-root-8"
>
  <h2
    class="MuiTypography-root MuiTypography-h4"
  >
    Cadastro de marcas
  </h2>
</div>
`;
